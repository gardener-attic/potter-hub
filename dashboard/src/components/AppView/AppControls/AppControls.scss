.upgrade-button {
  margin-right: 0.4em;
}

.notification {
  position: absolute;
  right: 15%;
  top: 22%;
}

.upgrade-text {
  padding-right: 1.25em;
  color: #008145;
}

.button:hover {
  .upgrade-text {
    color: inherit;
  }
  .notification {
    fill: transparent;
  }
}
