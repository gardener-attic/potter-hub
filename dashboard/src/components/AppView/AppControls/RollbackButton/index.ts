import RollbackButton from "./RollbackButton";

export default RollbackButton;
