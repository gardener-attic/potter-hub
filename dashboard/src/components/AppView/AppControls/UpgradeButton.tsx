import { RouterAction } from "connected-react-router";
import * as React from "react";
import { ArrowUpCircle } from "react-feather";

interface IUpgradeButtonProps {
  newVersion?: boolean;
  releaseName: string;
  releaseNamespace: string;
  push: (location: string) => RouterAction;
}

const UpgradeButton: React.SFC<IUpgradeButtonProps> = props => {
  const { newVersion, push, releaseName, releaseNamespace } = props;
  const onClick = () => push(`/apps/ns/${releaseNamespace}/upgrade/${releaseName}`);
  let upgradeButton = (
    <button className="button" onClick={onClick}>
      Upgrade
    </button>
  );
  // If the app is outdated highlight the upgrade button
  if (newVersion) {
    upgradeButton = (
      <button className="button upgrade-button" onClick={onClick}>
        <span className="upgrade-text">Upgrade</span>
        <ArrowUpCircle color="white" size={25} fill="#82C341" className="notification" />
      </button>
    );
  }
  return upgradeButton;
};

export default UpgradeButton;
