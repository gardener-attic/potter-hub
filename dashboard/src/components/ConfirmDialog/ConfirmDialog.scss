.confirm-dialog-loading-info {
  .loading-legend {
    line-height: 40px;
    padding-right: 5px;
  }

  .spinner {
    margin: 0;
  }
}
