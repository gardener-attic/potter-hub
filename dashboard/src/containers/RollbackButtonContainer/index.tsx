import RollbackButtonContainer from "./RollbackButtonContainer";

export default RollbackButtonContainer;
