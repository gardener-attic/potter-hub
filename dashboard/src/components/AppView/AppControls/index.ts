import AppControls from "./AppControls";

export default AppControls;
