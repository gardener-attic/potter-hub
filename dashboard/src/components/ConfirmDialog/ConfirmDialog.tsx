import * as React from "react";
import Modal from "react-modal";
import LoadingWrapper from "../LoadingWrapper";
import "./ConfirmDialog.css";

interface IConfirmDialogProps {
  modalIsOpen: boolean;
  loading: boolean;
  extraElem?: JSX.Element;
  confirmationText?: string;
  confirmationButtonText?: string;
  onConfirm: () => any;
  closeModal: () => any;
}

interface IConfirmDialogState {
  error?: string;
  modalIsOpen: boolean;
}

class ConfirmDialog extends React.Component<IConfirmDialogProps, IConfirmDialogState> {
  public state: IConfirmDialogState = {
    error: undefined,
    modalIsOpen: this.props.modalIsOpen,
  };

  public render() {
    return (
      <div className="ConfirmDialog">
        <Modal
          className="centered-modal"
          isOpen={this.props.modalIsOpen}
          onRequestClose={this.closeModal}
          contentLabel="Modal"
        >
          {this.state.error && (
            <div className="padding-big margin-b-big bg-action">{this.state.error}</div>
          )}
          {this.props.loading === true ? (
            <div className="row confirm-dialog-loading-info">
              <div className="col-8 loading-legend">Loading, please wait</div>
              <div className="col-4">
                <LoadingWrapper />
              </div>
            </div>
          ) : (
            <div>
              <div className="margin-b-normal">
                {this.props.confirmationText || "Are you sure you want to delete this?"}
              </div>
              {this.props.extraElem}
              <div className="margin-t-normal button-row">
                <button className="button" onClick={this.props.closeModal}>
                  Cancel
                </button>
                <button
                  className="button button-primary button-danger"
                  type="submit"
                  onClick={this.props.onConfirm}
                >
                  {this.props.confirmationButtonText || "Delete"}
                </button>
              </div>
            </div>
          )}
        </Modal>
      </div>
    );
  }

  public openModel = () => {
    this.setState({
      modalIsOpen: true,
    });
  };

  public closeModal = () => {
    this.setState({
      modalIsOpen: false,
    });
  };
}

export default ConfirmDialog;
