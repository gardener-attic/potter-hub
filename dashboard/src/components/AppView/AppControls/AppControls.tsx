import { RouterAction } from "connected-react-router";
import * as React from "react";
import { Redirect } from "react-router";

import { IRelease } from "shared/types";
import RollbackButtonContainer from "../../../containers/RollbackButtonContainer";
import ConfirmDialog from "../../ConfirmDialog";
import LoadingWrapper from "../../LoadingWrapper";
import "./AppControls.css";
import UpgradeButton from "./UpgradeButton";

interface IAppControlsProps {
  app: IRelease;
  deleteApp: (keepHistory: boolean) => Promise<boolean>;
  push: (location: string) => RouterAction;
}

interface IAppControlsState {
  migrate: boolean;
  modalIsOpen: boolean;
  redirectToAppList: boolean;
  upgrade: boolean;
  deleting: boolean;
  keepHistory: boolean;
}

class AppControls extends React.Component<IAppControlsProps, IAppControlsState> {
  public state: IAppControlsState = {
    deleting: false,
    migrate: false,
    modalIsOpen: false,
    keepHistory: false,
    redirectToAppList: false,
    upgrade: false,
  };

  public render() {
    const { app, push } = this.props;
    const { name, namespace } = app;
    const deleted = app.info && app.info.deleted;
    if (!name || !namespace) {
      return <LoadingWrapper />;
    }
    return (
      <div className="AppControls">
        {/* If the app has been deleted hide the upgrade button */}
        {!deleted && (
          <React.Fragment>
            <UpgradeButton
              newVersion={app.updateInfo && !app.updateInfo.upToDate}
              releaseName={name}
              releaseNamespace={namespace}
              push={push}
            />
            {/* We only show the rollback button if there are versions to rollback */}
            {app.version > 1 && (
              <RollbackButtonContainer
                releaseName={name}
                namespace={namespace}
                revision={app.version}
              />
            )}
          </React.Fragment>
        )}
        <button className="button button-danger" onClick={this.openModel}>
          {deleted ? "Purge" : "Delete"}
        </button>
        <ConfirmDialog
          onConfirm={this.handleDeleteClick}
          modalIsOpen={this.state.modalIsOpen}
          loading={this.state.deleting}
          closeModal={this.closeModal}
          extraElem={
            deleted ? (
              undefined
            ) : (
              <div className="margin-b-normal text-c">
                <label className="checkbox margin-r-big">
                  <input type="checkbox" onChange={this.togglePurge} />
                  <span>Keep History</span>
                </label>
              </div>
            )
          }
        />
        {this.state.redirectToAppList && <Redirect to={`/apps/ns/${namespace}`} />}
      </div>
    );
  }

  public openModel = () => {
    this.setState({
      modalIsOpen: true,
    });
  };

  public closeModal = async () => {
    this.setState({
      modalIsOpen: false,
    });
  };

  public handleDeleteClick = async () => {
    this.setState({ deleting: true });
    // Purge the release if the application has been already deleted
    const alreadyDeleted = this.props.app.info && !!this.props.app.info.deleted;
    const deleted = await this.props.deleteApp(!alreadyDeleted && this.state.keepHistory);
    const s: Partial<IAppControlsState> = { modalIsOpen: false };
    if (deleted) {
      s.redirectToAppList = true;
    }
    this.setState(s as IAppControlsState);
  };

  private togglePurge = () => {
    this.setState({ keepHistory: !this.state.keepHistory });
  };
}

export default AppControls;
