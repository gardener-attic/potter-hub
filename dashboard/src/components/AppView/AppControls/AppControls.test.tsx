import { mount, shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";
import Modal from "react-modal";
import { Redirect } from "react-router";
import RollbackButtonContainer from "../../../containers/RollbackButtonContainer";
import { hapi } from "../../../shared/hapi/release";
import itBehavesLike from "../../../shared/specs";
import ConfirmDialog from "../../ConfirmDialog";

import { IRelease } from "shared/types";
import AppControls from "./AppControls";
import UpgradeButton from "./UpgradeButton";

it("calls delete function when clicking the button", done => {
  const name = "foo";
  const namespace = "bar";
  const app = new hapi.release.Release({ name, namespace });
  const wrapper = shallow(
    <AppControls app={app} deleteApp={jest.fn(() => true)} push={jest.fn()} />,
  );
  const button = wrapper
    .find(".AppControls")
    .children()
    .find(".button-danger");
  expect(button.exists()).toBe(true);
  expect(button.text()).toBe("Delete");
  button.simulate("click");

  const confirm = wrapper
    .find(".AppControls")
    .children()
    .find(ConfirmDialog);
  expect(confirm.exists()).toBe(true);
  confirm.props().onConfirm(); // Simulate confirmation

  expect(wrapper.state("deleting")).toBe(true);
  // Wait for the async action to finish
  setTimeout(() => {
    wrapper.update();
    const redirect = wrapper.find(Redirect);
    expect(redirect.props()).toMatchObject({
      to: `/apps/ns/${namespace}`,
    } as any);
    done();
  }, 1);
});

it("calls delete function with additional purge", () => {
  const name = "foo";
  const namespace = "bar";
  const app = new hapi.release.Release({ name, namespace });
  const deleteApp = jest.fn(() => false); // Return "false" to avoid redirect when mounting
  // mount() is necessary to render the Modal
  const wrapper = mount(<AppControls app={app} deleteApp={deleteApp} push={jest.fn()} />);
  Modal.setAppElement(document.createElement("div"));
  wrapper.setState({ modalIsOpen: true });
  wrapper.update();

  // Check that the checkbox changes the AppControls state
  const confirm = wrapper.find(ConfirmDialog);
  expect(confirm.exists()).toBe(true);
  const checkbox = wrapper.find('input[type="checkbox"]');
  expect(checkbox.exists()).toBe(true);
  expect(wrapper.state("keepHistory")).toBe(false);
  checkbox.simulate("change");
  expect(wrapper.state("keepHistory")).toBe(true);

  // Check that the "purge" state is forwarded to deleteApp
  confirm.props().onConfirm(); // Simulate confirmation
  expect(deleteApp).toHaveBeenCalledWith(true);
});

context("when name or namespace do not exist", () => {
  const props = {
    app: new hapi.release.Release({ name: "name", namespace: "my-ns" }),
  };

  itBehavesLike("aLoadingComponent", {
    component: AppControls,
    props: { ...props, app: { ...props.app, name: null } },
  });
  itBehavesLike("aLoadingComponent", {
    component: AppControls,
    props: { ...props, app: { ...props.app, namespace: null } },
  });
});

context("when the application has been already deleted", () => {
  const props = {
    app: new hapi.release.Release({ name: "name", namespace: "my-ns", info: { deleted: {} } }),
    deleteApp: jest.fn(() => false), // Return "false" to avoid redirect when mounting
  };

  it("should show Purge instead of Delete in the button title", () => {
    const wrapper = shallow(<AppControls {...props} push={jest.fn()} />);
    const button = wrapper.find(".button-danger");
    expect(button.text()).toBe("Purge");
  });

  it("should not show the purge checkbox", () => {
    // mount() is necessary to render the Modal
    const wrapper = mount(<AppControls {...props} push={jest.fn()} />);
    Modal.setAppElement(document.createElement("div"));
    wrapper.setState({ modalIsOpen: true });
    wrapper.update();

    const confirm = wrapper.find(ConfirmDialog);
    expect(confirm.exists()).toBe(true);
    const checkbox = wrapper.find('input[type="checkbox"]');
    expect(checkbox).not.toExist();
  });

  it("should not keep history when status is deleted when clicking on delete", () => {
    // mount() is necessary to render the Modal
    const deleteApp = jest.fn(() => false);
    const wrapper = mount(<AppControls {...props} deleteApp={deleteApp} push={jest.fn()} />);
    Modal.setAppElement(document.createElement("div"));
    wrapper.setState({ modalIsOpen: true, keepHistory: true });
    wrapper.update();

    const confirm = wrapper.find(ConfirmDialog);
    expect(confirm.exists()).toBe(true);

    // Check that the "purge" is forwarded to deleteApp
    confirm.props().onConfirm(); // Simulate confirmation
    expect(deleteApp).toHaveBeenCalledWith(false);
  });

  it("should not show the Upgrade button", () => {
    const deleteApp = jest.fn(() => false);
    const wrapper = shallow(<AppControls {...props} deleteApp={deleteApp} push={jest.fn()} />);
    const buttons = wrapper.find("button");
    expect(buttons.length).toBe(1);
    expect(buttons.text()).toBe("Purge");
  });
});

context("when there is a new version available", () => {
  it("should forward the latest version", () => {
    const name = "foo";
    const namespace = "bar";
    const app = {
      name,
      namespace,
      updateInfo: {
        upToDate: false,
        chartLatestVersion: "1.0.0",
        appLatestVersion: "1.0.0",
      },
    } as IRelease;
    const wrapper = shallow(<AppControls app={app} deleteApp={jest.fn()} push={jest.fn()} />);

    expect(wrapper.find(UpgradeButton).prop("newVersion")).toBe(true);
  });
});

context("when the application is up to date", () => {
  it("should not forward the latest version", () => {
    const name = "foo";
    const namespace = "bar";
    const app = {
      name,
      namespace,
      updateInfo: {
        upToDate: true,
        chartLatestVersion: "1.1.0",
        appLatestVersion: "1.1.0",
      },
    } as IRelease;
    const wrapper = shallow(<AppControls app={app} deleteApp={jest.fn()} push={jest.fn()} />);

    expect(wrapper.find(UpgradeButton).prop("updateVersion")).toBe(undefined);
  });
});

context("Rollback button", () => {
  it("should show the RollbackButton when there is more than one revision", () => {
    const props = {
      app: new hapi.release.Release({
        name: "name",
        namespace: "my-ns",
        version: 2,
        info: {},
      }),
      deleteApp: jest.fn(() => false), // Return "false" to avoid redirect when mounting
    };
    const wrapper = shallow(<AppControls {...props} push={jest.fn()} />);
    const button = wrapper.find(RollbackButtonContainer);
    expect(button).toExist();
  });
  it("should not show the RollbackButton when there is only one revision", () => {
    const props = {
      app: new hapi.release.Release({
        name: "name",
        namespace: "my-ns",
        version: 1,
        info: {},
      }),
      deleteApp: jest.fn(() => false), // Return "false" to avoid redirect when mounting
    };
    const wrapper = shallow(<AppControls {...props} push={jest.fn()} />);
    const button = wrapper.find(RollbackButtonContainer);
    expect(button).not.toExist();
  });
});
