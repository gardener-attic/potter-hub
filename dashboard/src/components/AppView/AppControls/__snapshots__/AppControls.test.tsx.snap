// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when name or namespace do not exist loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
  type={0}
/>
`;

exports[`when name or namespace do not exist loading spinner matches the snapshot 2`] = `
<LoadingWrapper
  loaded={false}
  type={0}
/>
`;
